Index: gtk+3.0/gtk/ddefiledialog.c
===================================================================
--- gtk+3.0.orig/gtk/ddefiledialog.c
+++ gtk+3.0/gtk/ddefiledialog.c
@@ -461,6 +461,7 @@ static void _gtk_file_chooser_set_do_ove
 
 static void d_show_dbus_filedialog(GtkWidget *widget_ghost)
 {
+    guint64 dbus_dialog_winId;
     const gchar *title = gtk_window_get_title(GTK_WINDOW(widget_ghost));
     GtkFileChooserAction action = gtk_file_chooser_get_action(GTK_FILE_CHOOSER(widget_ghost));
 
@@ -484,12 +485,13 @@ static void d_show_dbus_filedialog(GtkWi
     if (!d_dbus_filedialog_call_by_ghost_widget_sync(widget_ghost, "show", NULL, NULL, NULL))
         return;
 
-    guint64 dbus_dialog_winId;
-
-    if (!d_dbus_filedialog_call_by_ghost_widget_sync(widget_ghost, "winId", NULL, "(t)", &dbus_dialog_winId))
-        return;
-
-    XSetTransientForHint(gdk_x11_get_default_xdisplay(), dbus_dialog_winId, GDK_WINDOW_XID(gtk_widget_get_window(widget_ghost)));
+    
+    if (GDK_IS_X11_DISPLAY(gtk_widget_get_display(widget_ghost))) {
+        if (!d_dbus_filedialog_call_by_ghost_widget_sync(widget_ghost, "winId", NULL, "(t)", &dbus_dialog_winId))
+            return;
+        XSetTransientForHint(gdk_x11_get_default_xdisplay(), dbus_dialog_winId, GDK_WINDOW_XID(gtk_widget_get_window(widget_ghost)));
+    }
+    
 }
 
 static void d_hide_dbus_filedialog(GtkWidget *widget_ghost)
@@ -919,7 +921,7 @@ static gboolean hook_gtk_file_chooser_di
     // 判断窗管协议类型，决定调用哪套dbus接口
     GdkDisplay *display = gtk_widget_get_display(dialog);
 #ifdef GDK_WINDOWING_WAYLAND
-    if (GDK_IS_WAYLAND_WINDOW(display)) {
+    if (GDK_IS_WAYLAND_DISPLAY(display)) {
         d_debug("dde-filedialog: gdk is wayland display!\n");
         if (wayland_dbus_is_exist()) {
             g_dbus_service = DFM_FILEDIALOG_DBUS_SERVER_WAYLAND;
